### ISSUE TITLE
Does this log the raw score?

### ISSUE THREAD

**Issue by dmadeka** [facebookresearch/nle#194]
In `agent.py`, if I were to modify https://github.com/facebookresearch/nle/blob/master/nle/agent/agent.py#L245

Adding the following lines:

```python
scores = batch['blstats'][:, :, 9][batch['done']]

stats = {
    "episode_returns": tuple(episode_returns.cpu().numpy()),
    "mean_episode_return": torch.mean(episode_returns).item(),
    "mean_score": torch.mean(torch.Tensor.float(scores)).item(),
    "total_loss": total_loss.item(),
    "pg_loss": pg_loss.item(),
    "baseline_loss": baseline_loss.item(),
    "entropy_loss": entropy_loss.item(),
}
```

Would that give me the raw score of the game (regardless of how I modified the episode reward)? Asking since it mostly has this output in the logfile:

```python
df.dropna(subset=['mean_score'])['mean_score'].describe()
count    15638.000000
mean         0.065729
std          0.437785
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max          7.000000
Name: mean_score, dtype: float64
```

**Comment by dmadeka** [comment]
Looking at the log files:

```
 'total_loss': 6000.3330078125}
[INFO:45159 agent:581 2021-07-08 23:15:52,635] Steps 894095360 @ 7672.2 SPS. Loss 15190.161133. Return per episode: 505.9. Stats:
{'baseline_loss': 15294.7431640625,
 'entropy_loss': -0.10292384773492813,
 'episode_returns': (576.36383, 663.9667, 277.32004),
 'episode_scores': (0, 0, 0),
 'mean_episode_return': 505.8835144042969,
 'mean_level': 1.0,
 'mean_score': 0.0,
 'pg_loss': -104.47933959960938,
 'total_loss': 15190.1611328125}
```

Doesnt really seem to add up - unclear why this wouldnt work

**Comment by dmadeka** [comment]
Are the scores reset at `done`? That might explain it - since the levels seem to always be 1 as well

**Comment by heiner** [comment]
Hey Dhruv,

Thanks for your interest in NLE!

I'm not 100% sure what the issue at hand is though? To answer some possibly related questions:

* Yes, the entry with index `9` of `blstats` is the score. This corresponds to the score as displayed in the bottom line stats during the game, highlighted here:
![image](https://user-images.githubusercontent.com/69371/125104945-d4eeae80-e0d5-11eb-8a10-8c01263d8666.png)
* The _reward_ the agent receives depends on the task, but one possible choice would be the turn-by-turn _difference_ of that score entry. Computing that requires saving the `blstats[9]` value from the last turn and comparing it with the current one (note that the hero in NetHack can lose score points).

**Comment by dmadeka** [comment]
From @ceisenach - one conjecture is that the `done` resets the `blstats`. So doing something like

```
def _get_scores(blstats, dones, score_index=9, level_index=24):
    mask = dones[1:, :]
    scores = blstats[:-1, :, score_index][mask]
    levels = blstats[:-1, :, level_index][mask]
    return scores, levels
```

Would get me the scores and levels from the end of the episode. The reason I pasted this was that the stats for the scores and levels didnt match what I was seeing from the `ttyrec` files (where the average score was ~600) but the max score in the `logs.tsv` file seems to be 7 (shown above)

**Comment by tscmoo** [comment]
Indeed, in the file you linked, at https://github.com/facebookresearch/nle/blob/master/nle/agent/agent.py#L367, when `done`, the obs returned are those from `reset`, which would be the start of a new episode. This would of course include the blstats, so when `done`, you get the `blstats` if the new game, not the one that just finished.

**Comment by heiner** [comment]
Hey Dhruv,

Please re-open in case this doesn't answer your question.

**Comment by dmadeka** [comment]
Yup - part of this was just to keep a googleable search for the `done` question :)

**Comment by heiner** [comment]
Generally, in IMPALA and related environments, the final observation of an episode can be discarded as no learning from it can happen (no more actions get applied afterwards). Hence, both monobeast and polybeast replace that observation with the first observation from the next episode. That way, one has an infinite stream of data coming from the environment, which then includes a boolean `done` in each step.

I believe OpenAI have experimented with similar designs for their [`gym3` API](https://github.com/openai/gym3/blob/master/docs/design.md).

In the case of NetHack, I realize though there's a few interesting consequences: As we skip over end-of-game frames, the final value of the score might not be easily accessible if it's different from the final reward (which it would be when reward shaping along the lines of "disincentivizing not changing the time counter" is used). Moreover, in a sense there's still _information_ on that final frame: E.g., which of potentially many lurking dangers ended up killing the hero. I know of no RL approach that could make use of that information, but it is potentially usable.

**Comment by dmadeka** [comment]
Actually it was exactly to use that information!
